### PR TITLE
CLI Refactor: Seclude service configuration path resolution out of internals

### PR DESCRIPTION
**Pull request by medikoo**
Addresses 2.0 from https://github.com/serverless/serverless/issues/8364

In a side PR  (https://github.com/serverless/test/pull/69) it's ensured that `runServerless` adapts to new way of serverless config path resolution and passes  `configurationPath` to `Serverless` constructor (still it's not mandatory for successful test runs)

Additionally (all independent updates where submitted with individual commits)
- Apply async/await refactor to touched parts
- Fix issues in interactive new project setup:
   - Ensure to reject setup if proposed new project path already exists (so far we rejected only if provided path pointed the actual Serverless service, which was definitely error prone as that was followed by template dir copy)
   - Ensure internally updated `Serverless` instance has properly setup config references (this bug probably didn't introduce any visible issues)
- Ensure that memoized `getServerlessConfigFile` is treated as _async_ function (so cache is not stored for rejected value)

### CODE REVIEW

File: lib/utils/getServerlessConfigFile.js [hunk 1]
```diff
@@ -108,23 +63,24 @@ const handleJsOrTsConfigFile = (configFile) =>
...
+    throw new ServerlessError(
+      `${path.basename(configFilePath)} must export plain object`,
+      'INVALID_CONFIG_OBJECT_TYPE'
+    );
+  },
+  { promise: true }
 );
 
-module.exports = { getConfigFilePath, getServerlessConfigFile, getServerlessConfigFilePath };
+module.exports = { getServerlessConfigFile };
```
Comment (pgrzesik) [inline lib/utils/getServerlessConfigFile.js:86]: Any particular reason why it's not exporting it directly with `module.exports = getServerlessConfigFile` ?

**Review by pgrzesik — APPROVED** [review]
I really tried, but I didn't find any suggestions other than one "out of curiosity" question. Well done :tada:

File: lib/utils/getServerlessConfigFile.js (hunk 1 above)
Comment (medikoo) [reply lib/utils/getServerlessConfigFile.js:86]: Good point. Technically it exported many values, and I just removed some, it didn't bother me it exports just one, as anyway this whole util is scheduled to be removed with next step (In light of that I don't think there's a point in cleaning this up further)

Comment (pgrzesik) [reply lib/utils/getServerlessConfigFile.js:86]: Sounds good, I was just curious :+1:

File: lib/cli/resolve-configuration-path.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    const configParamIndex = (() => {
+      const paramIndex = process.argv.indexOf('--config');
+      if (paramIndex !== -1) return paramIndex;
+      return process.argv.indexOf('-c');
+    })();
+    return configParamIndex === -1 ? null : process.argv[configParamIndex + 1];
+  })();
+
+  if (customConfigName) {
+    if (path.isAbsolute(customConfigName)) {
```
Comment (ankon) [inline lib/cli/resolve-configuration-path.js:21]: Is there any reason why a configuration path cannot be absolute? You already handle that nicely by using `path.resolve` just below (compared to the previous `path.join`)?

After all, this just makes it ever so slightly harder: Now I need to take my absolute path, and relativize it against the current directory, and then provide that ...

Comment (medikoo) [reply lib/cli/resolve-configuration-path.js:21]: @ankon Thanks for that comment, that's a very valid point. I've prepared a fix at: https://github.com/serverless/serverless/pull/8793

**Comment by jgeurts** [comment]
This caused a breaking change for us, unfortunately, by preventing configs in sub-directories.

**Comment by medikoo** [comment]
@jgeurts please jump in to #8800

**Pull request merged**